[BREAKING]libc/elf: rename modlib to libelf

anchao opened the pull request:
## Summary

libc/elf: rename modlib to libelf

Renaming "modlib" to "elf" is more in line with the implementation content,
which makes it easier for individual developers to understand the capabilities of this module.

CONFIG_LIBC_MODLIB -> CONFIG_LIBC_ELF

Signed-off-by: chao an <anchao.archer@bytedance.com>

## Impact

N/A, depends on https://github.com/apache/nuttx-apps/pull/2985

## Testing

sim/sotest

nuttxpr commented:
[**\[Experimental Bot, please feedback here\]**](https://github.com/search?q=repo%3Aapache%2Fnuttx+13552&type=issues)



No. This PR, as described, does not fully meet the NuttX requirements.  Here's why:

* **Insufficient Impact Assessment:**  Simply stating "N/A" for impact is unacceptable.  Even a seemingly simple rename can have impacts.  At a minimum, the developer needs to confirm:
    * **Impact on user:** Will any application code need to be changed due to the config option rename?  Even if the answer is no, it needs to be explicitly stated.
    * **Impact on documentation:** The documentation needs to be updated to reflect the name change.  The PR should either include these changes or explain why they are not necessary.
    * **Impact on compatibility:** Does this change break any backward compatibility? Again, even if the answer is no, it needs to be explicitly stated and justified.

* **Insufficient Testing Information:**  "sim/sotest" is not enough information.  The PR needs to specify:
    * **Build Host:**  What operating system, compiler version, etc. was used to build NuttX?
    * **Target Details:** What simulator was used (qemu, etc.)? What architecture? What configuration?
    * **Testing Logs:**  The PR includes placeholders for logs, but no actual logs are provided.  There needs to be concrete evidence that the change works as intended.  What tests were run? What were the results before and after the change?


In short, while the summary provides a reasonable explanation for the *why*, the impact and testing sections lack crucial details required for proper review and acceptance.

lupyuen reviewed (changes requested):
Sorry @anchao I tested the new branch `anchao:25020501` for NuttX and Apps, it fails to build for rv-virt:knsh64. Did I miss something? Thanks :-)
https://gist.github.com/lupyuen/58c3b955a0c222776af523662561c494
```text
+ git clone https://github.com/anchao/nuttx --branch 25020501 nuttx
+ git clone https://github.com/anchao/nuttx-apps --branch 25020501 apps
+ tools/configure.sh rv-virt:knsh64
+ make -j

CC:  driver/fs_registerblockdriver.c In file included from binfmt_globals.c:29:
binfmt.h:261:5: error: conflicting types for 'elf_initialize'; have 'int(void)'
  261 | int elf_initialize(void);
      |     ^~~~~~~~~~~~~~
```

xiaoxiang781216 commented:
> Sorry @anchao I tested the new branch `anchao:25020501` for NuttX and Apps, it fails to build for rv-virt:knsh64. Did I miss something? Thanks :-) https://gist.github.com/lupyuen/58c3b955a0c222776af523662561c494
> 
> ```
> + git clone https://github.com/anchao/nuttx --branch 25020501 nuttx
> + git clone https://github.com/anchao/nuttx-apps --branch 25020501 apps
> + tools/configure.sh rv-virt:knsh64
> + make -j
> 
> CC:  driver/fs_registerblockdriver.c In file included from binfmt_globals.c:29:
> binfmt.h:261:5: error: conflicting types for 'elf_initialize'; have 'int(void)'
>   261 | int elf_initialize(void);
>       |     ^~~~~~~~~~~~~~
> ```

@lupyuen can the build bot copy the this info automatically? so, we can hold the change until the break get fixed.

anchao commented:
> Sorry @anchao I tested the new branch `anchao:25020501` for NuttX and Apps, it fails to build for rv-virt:knsh64. Did I miss something? Thanks :-) https://gist.github.com/lupyuen/58c3b955a0c222776af523662561c494
> 
> ```
> + git clone https://github.com/anchao/nuttx --branch 25020501 nuttx
> + git clone https://github.com/anchao/nuttx-apps --branch 25020501 apps
> + tools/configure.sh rv-virt:knsh64
> + make -j
> 
> CC:  driver/fs_registerblockdriver.c In file included from binfmt_globals.c:29:
> binfmt.h:261:5: error: conflicting types for 'elf_initialize'; have 'int(void)'
>   261 | int elf_initialize(void);
>       |     ^~~~~~~~~~~~~~
> ```

@lupyuen Thank you, I updated my modification and changed elf to libelf, so that rv-virt:knsh64 can be compiled successfully, but I found a new issue, rv-virt:knsh64 cannot be started normally on my platform, even if I roll back my changes

anchao commented:
> @lupyuen Thank you, I updated my modification and changed elf to libelf, so that rv-virt:knsh64 can be compiled successfully, but I found a new issue, rv-virt:knsh64 cannot be started normally on my platform, even if I roll back my changes

@lupyuen I just asked @yf13 to help verify this PR, which could be bringup normally on qemu6.2. My qemu version is 8.2.2. It seems that there are some compatibility issues

lupyuen commented:
Cool lemme know when we're ready to retest, I'll run my script again. Thanks :-)

anchao commented:
> Cool lemme know when we're ready to retest, I'll run my script again. Thanks :-)

It seems that qemu 8.2.2 cannot execute rv-virt/knsh64 not caused by my PR. If your test environment is lower than this version, please restart your script. Thank you.

xiaoxiang781216 commented on Documentation/applications/examples/module/index.rst:
```diff
@@ -55,8 +55,8 @@ configuration options:
 6. Linker scripts. You might also want to use a linker scripts to combine
    sections better. An example linker script is at
-   ``nuttx/libc/modlib/gnu-elf.ld``. That example might have to be tuned for your
+   ``nuttx/libc/libelf/gnu-elf.ld``. That example might have to be tuned for your
```
could we name it elf?
```
   ``nuttx/libc/elf/gnu-elf.ld``. That example might have to be tuned for your
```

anchao replied:
done

xiaoxiang781216 commented on Documentation/platforms/arm/samv7/boards/samv71-xult/README.txt:
```diff
@@ -1819,10 +1819,10 @@ Configuration sub-directories
-         CONFIG_MODLIB_BUFFERINCR=32
-         CONFIG_MODLIB_BUFFERSIZE=128
+         CONFIG_LIBC_LIBELF=y
+         CONFIG_LIBELF_ALIGN_LOG2=2
```
```suggestion
         CONFIG_LIBc_ELF_ALIGN_LOG2=2
```

anchao replied:
done

xiaoxiang781216 commented on Documentation/platforms/arm/samv7/boards/samv71-xult/README.txt:
```diff
@@ -1819,10 +1819,10 @@ Configuration sub-directories
-         CONFIG_MODLIB_ALIGN_LOG2=2
-         CONFIG_MODLIB_BUFFERINCR=32
-         CONFIG_MODLIB_BUFFERSIZE=128
+         CONFIG_LIBC_LIBELF=y
```
```suggestion
         CONFIG_LIBC_ELF=y
```

xiaoxiang781216 commented on libs/libc/libelf/libelf.h:
```diff
@@ -1,5 +1,5 @@
 /****************************************************************************
- * libs/libc/modlib/modlib.h
+ * libs/libc/libelf/libelf.h
```
libelf_*.[c|h] to elf_*.[c|h]

anchao replied:
done

anchao commented:
> > Cool lemme know when we're ready to retest, I'll run my script again. Thanks :-)
> 
> It seems that qemu 8.2.2 cannot execute rv-virt/knsh64 not caused by my PR. If your test environment is lower than this version, please restart your script. Thank you.

@lupyuen Sorry to bother you again. I've updated the naming of some relevant macros and code directories. we may need to run the script again.

lupyuen reviewed (approved):
Tested OK on rv-virt:knsh64. Thanks :-)
https://gist.github.com/lupyuen/edb4747d1235c77be624bd4c6a24a500
```text
+ git clone https://github.com/anchao/nuttx --branch 25020501 nuttx
+ git clone https://github.com/anchao/nuttx-apps --branch 25020501 apps
NuttX Source: https://github.com/apache/nuttx/tree/fa059c19fad275324afdfec023d24a85827516e9
NuttX Apps: https://github.com/apache/nuttx-apps/tree/6d0afa6c9b8d4ecb896f9aa177dbdfcd40218f48
+ tools/configure.sh rv-virt:knsh64
nsh> uname -a
NuttX 10.4.0 fa059c19fa Feb  5 2025 19:25:45 risc-v rv-virt
nsh> ostest
ostest_main: Exiting with status 0
```

lupyuen reviewed (approved):
Also tested OK on real hardware: Oz64 SG2000 RISC-V SBC (`milkv_duos:nsh`). Thanks :-)
https://gist.github.com/lupyuen/15fd5e68f66b03f19f476b5ef3c066f1
```text
+ git clone https://github.com/anchao/nuttx nuttx --branch 25020501
+ git clone https://github.com/anchao/nuttx-apps apps --branch 25020501
NuttX Source: https://github.com/apache/nuttx/tree/fa059c19fad275324afdfec023d24a85827516e9
NuttX Apps: https://github.com/apache/nuttx-apps/tree/6d0afa6c9b8d4ecb896f9aa177dbdfcd40218f48
+ tools/configure.sh milkv_duos:nsh
nsh> ostest
ostest_main: Exiting with status 0
```

cederom reviewed (changes requested):
Thanks @anchao :-)
* Why this change is really necessary? What it fixes / improves?
* Modlib library [1] is meant for loadable modules of various formats. So far we only support ELF but other formats may be added. When other formats are added then elf name will be confusing again. We should stick to name modlib as this is loadable module component, not just elf handler.
* How does that change impact existing configurations and backward-compatibility (i.e. older nuttx-apps)? You marked impact `N/A` but this seems breaking change so description is invalid? Also providing reference to different change is not an impact factor just a reference / dependency and should belong to summary section.
* This change will breaks existing libc and configuration. It will break compatibility with nuttx-apps out of sync with nuttx. This change should be marked as breaking change. It would be good to first discuss this kind of change on mailing lists.
* Where are build / runtime test logs? Did you verify build and runtime with latest release of nuttx-apps to see if it does not break things?
* How does that align with other standards (libc) and OS/RTOS implementations?
* Okay I can see that `modlib_` calls  [2] are renamed to `libelf_`. Is this in align and compatibility with existing `libelf` library?

[1] https://nuttx.apache.org/docs/latest/applications/examples/module/index.html
[2] https://github.com/apache/nuttx/tree/fdc0b608b5893a207ea37a276e0adade6ded3b62/libs/libc/modlib

anchao commented:
> Thanks @anchao :-)
> 
> * Why this change is really necessary? What it fixes / improves?
> * Modlib library [1] is meant for loadable modules of various formats. So far we only support ELF but other formats may be added. When other formats are added then elf name will be confusing again. We should stick to name modlib as this is loadable module component, not just elf handler.
> * How does that change impact existing configurations and backward-compatibility (i.e. older nuttx-apps)? You marked impact `N/A` but this seems breaking change so description is invalid? Also providing reference to different change is not an impact factor just a reference / dependency and should belong to summary section.
> * This change will breaks existing libc and configuration. It will break compatibility with nuttx-apps out of sync with nuttx. This change should be marked as breaking change. It would be good to first discuss this kind of change on mailing lists.
> * Where are build / runtime test logs? Did you verify build and runtime with latest release of nuttx-apps to see if it does not break things?
> * How does that align with other standards (libc) and OS/RTOS implementations?
> * Okay I can see that `modlib_` calls  [2] are renamed to `libelf_`. Is this in align and compatibility with existing `libelf` library?
> 
> [1] https://nuttx.apache.org/docs/latest/applications/examples/module/index.html [2] https://github.com/apache/nuttx/tree/fdc0b608b5893a207ea37a276e0adade6ded3b62/libs/libc/modlib

https://github.com/apache/nuttx-apps/pull/2985#issuecomment-2641700861

pussuw commented:
I know about the implementation duplication before, libelf and modlib were about the same. However, will we have a need to "re-introduce" modlib, if it gets capability to load modules other than elf format ? Or will we just implement the new module loader into whateverlib and re-create modlib as a wrapper ?

Because I suspect the original intention of modlib was not at all to duplicate code, it was supposed to be higher in the code hierarchy compared to libelf ?

anchao commented:
> I know about the implementation duplication before, libelf and modlib were about the same. However, will we have a need to "re-introduce" modlib, if it gets capability to load modules other than elf format ? Or will we just implement the new module loader into whateverlib and re-create modlib as a wrapper ?
> 
> Because I suspect the original intention of modlib was not at all to duplicate code, it was supposed to be higher in the code hierarchy compared to libelf ?

Actually, that's not the case. In the architectural design, the structure of binfmt should be reused to complete the glue code. modlib is just a part of libelf. If we consider the design from the user's perspective, some future loadable types, such as coff, fdpic, flat, misc, script, etc., should be integrated with binfmt rather than with modlib.

https://github.com/torvalds/linux/blob/master/fs/binfmt_elf.c
https://github.com/torvalds/linux/blob/master/fs/binfmt_elf_fdpic.c
https://github.com/torvalds/linux/blob/master/fs/binfmt_flat.c
https://github.com/torvalds/linux/blob/master/fs/binfmt_misc.c
https://github.com/torvalds/linux/blob/master/fs/binfmt_script.c
https://github.com/alexbousso/kernel_2.4.18-14/blob/master/fs/binfmt_coff.c

This is where they belong:
https://github.com/apache/nuttx/tree/master/binfmt

anchao commented:
I'm not saying that the structure of Linux must be correct. However, I think that @gregory-nutt  and Linus had similar ideas in the initial design. Compared with the kernels implements of the other 3-parties, I believe that the initial design of NuttX is more reasonable.

pussuw reviewed (approved):
Thanks for the explanation, LGTM!

jerpelea commented:
@anchao please rebase

acassis commented:
@anchao please verify the compilation issue

cederom commented:
* Okay, there seems to be a common agreement to approve this change, so I will remove my blocking vote :-)
* Do we want to think of compatibility with existing libelf implementations here? :-)

anchao commented:
> * Okay, there seems to be a common agreement to approve this change, so I will remove my blocking vote :-)

Thank you. :)

> * Do we want to think of compatibility with existing libelf implementations here? :-)

Yes, this commit is just naming change, I think the impact is under control.

anchao commented:
> @anchao please verify the compilation issue

Thanks @acassis , this PR depends on apps repo https://github.com/apache/nuttx-apps/pull/2985, so it may not be possible to pass CI separately

cederom commented:
* Do we want to tag this change with `[BREAKING]` tag in the PR titile and git commits? 
* It will impact build of nuttx/nuttx-apps and whoever used modlib.
* Maybe a candidate for future breaking 13 release? :-P

anchao commented:
> * Do we want to tag this change with `[BREAKING]` tag in the PR titile and git commits?

Done.

> * It will impact build of nuttx/nuttx-apps and whoever used modlib.

The linkage changes of nuttx/apps is normal case. You can check the recent changes of LINE_MAX/PATH_MAX.

> * Maybe a candidate for future breaking 13 release? :-P

Fine to me

anchao commented:
@cederom so do we have the plan to merge this PR?

lupyuen commented:
If we decide to merge this PR: Avaota-A1 needs to be updated too (`modlib` becomes `elf`)
https://github.com/apache/nuttx/blob/cf6781efade76f7f369673d7e83c3f9728de408e/boards/arm64/a527/avaota-a1/scripts/Make.defs#L52

anchao commented:
> If we decide to merge this PR: Avaota-A1 needs to be updated too (`modlib` becomes `elf`)
> 
> https://github.com/apache/nuttx/blob/cf6781efade76f7f369673d7e83c3f9728de408e/boards/arm64/a527/avaota-a1/scripts/Make.defs#L52

One question, if I fix the incremental changes of Avaota-A1, could this PR be merged? I have other features that depend on this change.

lupyuen commented:
@cederom: Wonder if we should wait for the NuttX Release to be cleared first, before merging this PR? Otherwise the backporting of fixes might get messy 🤔

anchao commented:
> @cederom: Wonder if we should wait for the NuttX Release to be cleared first, before merging this PR? Otherwise the backporting of fixes might get messy 🤔

I noticed that 12.9 has already pulled the release branch, why can't I merge this PR? Isn't it included in the 13.0 release?

anchao commented:
> If we decide to merge this PR: Avaota-A1 needs to be updated too (`modlib` becomes `elf`)
> 
> https://github.com/apache/nuttx/blob/cf6781efade76f7f369673d7e83c3f9728de408e/boards/arm64/a527/avaota-a1/scripts/Make.defs#L52

Done

lupyuen commented:
Tested OK with OSTest on Avaota-A1 Arm64 SBC. Thanks :-)
https://gist.github.com/lupyuen/8c5a77348559098818cbc72002225ef0
```bash
+ git clone https://github.com/anchao/nuttx nuttx --branch 25020501
+ git clone https://github.com/anchao/nuttx-apps apps --branch 25020501
+ tools/configure.sh avaota-a1:nsh
nsh> uname -a
NuttX 10.4.0 9a4b65d638 Apr  3 2025 20:43:22 arm64 avaota-a1
nsh> ostest
ostest_main: Exiting with status 0
```

lupyuen commented:
Since the CI Jobs won't run correctly for this Breaking PR: I'm running the CI Jobs myself through Docker. The CI Logs will appear here:
- https://gist.github.com/search?q=user%3Anuttxpr+run-ci-special&ref=searchresults

I'm running a script that calls Docker for all CI Jobs:
- https://github.com/lupyuen/nuttx-release/blob/main/run-ci-special.sh
```bash
  for job in \
    arm-01 arm-02 arm-03 arm-04 \
    arm-05 arm-06 arm-07 arm-08 \
    arm-09 arm-10 arm-11 arm-12 \
    arm-13 arm-14 \
    arm64-01 \
    other \
    risc-v-01 risc-v-02 risc-v-03 risc-v-04 \
    risc-v-05 risc-v-06 risc-v-07 \
    sim-01 sim-02 sim-03 \
    x86_64-01 \
    xtensa-01 xtensa-02 xtensa-03
  do

    sudo docker run -it \
      ghcr.io/apache/nuttx/apache-nuttx-ci-linux:latest \
      /bin/bash -c "
      set -x ;
      uname -a ;
      cd ;
      pwd ;
      git clone https://github.com/anchao/nuttx nuttx --branch 25020501 ;
      git clone https://github.com/anchao/nuttx-apps apps --branch 25020501 ;
      pushd nuttx ; git reset --hard HEAD ; echo NuttX Source: https://github.com/apache/nuttx/tree/\$(git rev-parse HEAD) ; popd ;
      pushd apps  ; git reset --hard HEAD ; echo NuttX Apps: https://github.com/apache/nuttx-apps/tree/\$(git rev-parse HEAD) ; popd ;
      cd nuttx/tools/ci ;
      ( sleep 10800 ; echo Killing pytest after timeout... ; pkill -f pytest )&
      (./cibuild.sh -c -A -N -R testlist/$job.dat || echo '***** BUILD FAILED') ;    
      "
  done
```

lupyuen reviewed (changes requested):
Sorry @anchao: The build for `qemu-armv7a:full` is failing. Should we rebase to master the PR for NuttX Apps?
https://gist.github.com/nuttxpr/d38255176bc9e55ff15bbcc2c86b8f7f#file-ci-arm-05-log-L1306
```bash
sudo docker run \
  -it \
  ghcr.io/apache/nuttx/apache-nuttx-ci-linux:latest \
  /bin/bash
cd
git clone https://github.com/anchao/nuttx nuttx --branch 25020501
git clone https://github.com/anchao/nuttx-apps apps --branch 25020501
cd nuttx/tools/ci
./cibuild.sh -c -A -N -R testlist/arm-05.dat

Cmake in present: qemu-armv7a/full,CONFIG_ARM_TOOLCHAIN_GNU_EABI
CMake Error at cmake/nuttx_add_library.cmake:190 (add_library):
  Cannot find source file:
    /root/apps/interpreters/quickjs/quickjs/libbf.c
Call Stack (most recent call first):
  /root/apps/interpreters/quickjs/CMakeLists.txt:127 (nuttx_add_library)
CMake Error at cmake/nuttx_add_library.cmake:190 (add_library):
  No SOURCES given to target: libqjs
Call Stack (most recent call first):
  /root/apps/interpreters/quickjs/CMakeLists.txt:127 (nuttx_add_library)
```

lupyuen reviewed (changes requested):
Hi @anchao: NuttX Build Farm has reported 5 Failed Builds. Maybe rebasing the NuttX Apps PR will help?

~~(1) qemu-armv7a/full~~
~~https://gist.github.com/nuttxpr/d38255176bc9e55ff15bbcc2c86b8f7f#file-ci-arm-05-log-L1307~~
```
Cannot find source file:
/root/apps/interpreters/quickjs/quickjs/libbf.c
```
Update: qemu-armv7a/full is now successful: https://gist.github.com/nuttxpr/91b660a47afdf6294d117a003fcf51b4

~~(2) sim/citest~~
~~https://gist.github.com/nuttxpr/c24146f4014353710306b87dd37b0f2c#file-ci-sim-01-log-L252~~
```
test_framework/test_cmocka.py::test_cmocka FAILED                                            [  0%]
test_example/test_example.py::test_hello FAILED                                              [  0%]
test_example/test_example.py::test_helloxx FAILED                                            [  0%]
test_example/test_example.py::test_pipe FAILED                                               [  0%]
4 failed, 1054 passed
```
Update: sim/citest is now successful: https://gist.github.com/lupyuen/a6f418105def27c13b86e53bb48967ea

~~(3) sim/quickjs~~
~~https://gist.github.com/nuttxpr/e4bdeb204df308f80c82463630eebe29#file-ci-sim-03-log-L226~~
```
Cannot find source file:
/root/apps/interpreters/quickjs/quickjs/libbf.c
```
Update: sim/quickjs is now successful: https://gist.github.com/nuttxpr/5e68d18e83062f852260cfa85ba020ea

~~(4) esp32-devkitc/snake~~
~~https://gist.github.com/nuttxpr/630c4e2f52dea96edc68fdb5fa1d9e10#file-ci-xtensa-01-log-L920~~
```
  [1/1] Normalize esp32-devkitc/snake
33,34d32
< CONFIG_GAMES_SNAKE=y
< CONFIG_GAMES_SNAKE_USE_GPIO=y
```
Update: esp32-devkitc/snake is now successful: https://gist.github.com/nuttxpr/b2fce9c35dc98772d5a5ccaeae209bee

(5) esp32s3-devkit/python
https://gist.github.com/nuttxpr/4dfc52fc3b42450cee9b7dbe5409f408#file-ci-xtensa-03-log-L169
```
make[3]: *** /root/apps/testing/sched/cpuload: No such file or directory.  Stop.
```

[Complete list of Build Failures in NuttX Dashboard](https://nuttx-dashboard.org/d/fe2bqg6uk7nr4a/build-logs-from-nuttx-build-farm?from=now-2d&to=now&timezone=browser&var-arch=$__all&var-subarch=$__all&var-board=$__all&var-config=$__all&var-group=$__all&var-Filters=user%7C%3D%7Cnuttxpr)

anchao commented:
> test_helloxx

@lupyuen Thanks a lot for your verification. I just rebased the changes under apps repo and tested the following configurations,  compile and unit test are all passed, could you help trigger the local test again?

(1) qemu-armv7a/full
(2) sim/citest
(3) sim/quickjs
(4) esp32-devkitc/snake
(5) esp32s3-devkit/python

anchao commented:
> Thanks @anchao! NuttX Build Farm is already building all jobs in a loop. We'll find out in 36 hours if all the builds are successful. Then I'll merge the NuttX Apps PR, followed by the NuttX PR. Thanks :-)

@lupyuen Thank you so much.

lupyuen commented:
Sorry @tmedicci: Is the build for `esp32s3-devkit:python` successful when we see these messages for our Special CI Build?
https://gist.github.com/nuttxpr/ea0bcb70eec1207d7b83290535d4ea34#file-ci-xtensa-03-log-L183-L233
```
configure: WARNING: no system libmpdecimal found; falling back to bundled libmpdecimal (deprecated and scheduled for removal in Python 3.15)
  WARNING: The scripts pip3 and pip3.13 are installed in '/root/apps/interpreters/python/install/host/bin' which is not on PATH.
  Consider adding this directory to PATH or, if you prefer to suppress this warning, use --no-warn-script-location.
WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager, possibly rendering your system unusable.It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv. Use the --root-user-action option if you know what you are doing and want to suppress this warning.
configure: WARNING: using cross tools not prefixed with host triplet
configure: WARNING: xtensa-esp32s3-nuttx/gcc is not supported
/root/apps/interpreters/python/Python/configure: line 10553: test: too many arguments
configure: WARNING: no system libmpdecimal found; falling back to bundled libmpdecimal (deprecated and scheduled for removal in Python 3.15)
/root/apps/interpreters/python/Python/configure: line 27327: test: =: unary operator expected
configure: WARNING:
Platform "xtensa-esp32s3-nuttx" with compiler "gcc" is not supported by the
CPython core team, see https://peps.python.org/pep-0011/ for more information.
Created /root/apps/interpreters/python/build/target/lib/python313.zip (3.11 MiB)
  [1/1] Normalize esp32s3-devkit/python
Untracked files:
	libs/libc/elf_sys_symtab.c
```

The above messages have appeared since March 30 (probably earlier), so I assume they are OK? `testbuild.sh` is terminating with an error, probably because of the Untracked File. Thanks!
- https://gist.github.com/nuttxpr/556dd4a808060ed0ea9687888cbbb755#file-ci-xtensa-03-log-L179-L229

anchao commented:
> Hi @anchao: The CI Checks by NuttX Build Farm will complete soon, so far no problems. [As mentioned](https://github.com/apache/nuttx-apps/pull/2985#pullrequestreview-2745192979) by @hartmannathan, could you please broadcast the Breaking Change to the mailing list?
> 
> > Could there be an email to the mailing list to point out the (proposed) breaking change? Some members of the community don't monitor GitHub. Thanks!
> 
> We should follow the process for Breaking Changes. Thanks :-)
> 
> * https://github.com/apache/nuttx/blob/master/CONTRIBUTING.md#113-breaking-changes-handling-process

@lupyuen  Done, thanks for reminder, I have sent an email to the dev list

yamt commented:
> Hi @anchao: The CI Checks by NuttX Build Farm will complete soon, so far no problems. [As mentioned](https://github.com/apache/nuttx-apps/pull/2985#pullrequestreview-2745192979) by @hartmannathan, could you please broadcast the Breaking Change to the mailing list?
> 
> > Could there be an email to the mailing list to point out the (proposed) breaking change? Some members of the community don't monitor GitHub. Thanks!
> 
> We should follow the process for Breaking Changes. Thanks :-)
> 
>     * https://github.com/apache/nuttx/blob/master/CONTRIBUTING.md#113-breaking-changes-handling-process

although i personally like this PR,
the change looks far from "absolutely necessary and unavoidable" to me.
what do you think?
maybe the wording in CONTRIBUTING.md is too strong?

lupyuen reviewed (changes requested):
Hi @anchao: NuttX Build Farm has successfully completed the CI Checks for this PR:
- [CI Logs are here](https://gist.github.com/search?q=8ab4db61a822c0093bb2fc5dddc3d75886f686f1)
- We restarted two CI Jobs due to failed library downloads and intermittent CI Test errors: [arm-05](https://gist.github.com/lupyuen/e52f7b11fcdf518862405221605c32e9) and [sim-01](https://gist.github.com/lupyuen/a6f418105def27c13b86e53bb48967ea)
- We'll ignore the [messages for esp32s3-devkit:python](https://gist.github.com/lupyuen/86e763ea742eac361d73e40a4f9b39cd#file-gistfile1-txt-L115-L167), since the messages are the [same before the PR](https://github.com/apache/nuttx/pull/15765#issuecomment-2782468664)

I'll wait for you to Close the Voting, and announce the decision to Merge the PR. Then I'll merge the NuttX Apps PR, followed by the NuttX PR. Thanks :-)

lupyuen commented:
Hi @anchao: The voting will close in 24 hours and we have (I think) [One Blocking Vote](https://lists.apache.org/thread/v7d7k1n2d1tl8n6o3pnocbz33knx3gmp).

Perhaps you could clarify via the Mailing List: Is Yamamoto-san objecting to this Breaking PR because it shouldn't be merged? Or is there an objection to the [NuttX Breaking Changes Handling Process](https://github.com/apache/nuttx/blob/master/CONTRIBUTING.md#113-breaking-changes-handling-process)?

If there are objections on the Breaking Changes Process: Maybe we can follow-up with [a separate discussion](https://github.com/apache/nuttx/pull/15950), so we don't block the Breaking PR. Thanks :-)

anchao commented:
> Hi @anchao: The voting will close in 24 hours and we have (I think) [One Blocking Vote](https://lists.apache.org/thread/v7d7k1n2d1tl8n6o3pnocbz33knx3gmp).
> 
> Perhaps you could clarify via the Mailing List: Is Yamamoto-san objecting to this Breaking PR because it shouldn't be merged? Or is there an objection to the [NuttX Breaking Changes Handling Process](https://github.com/apache/nuttx/blob/master/CONTRIBUTING.md#113-breaking-changes-handling-process)?
> 
> If there are objections on the Breaking Changes Process: Maybe we can follow-up with [a separate discussion](https://github.com/apache/nuttx/pull/15950), so we don't block the Breaking PR. Thanks :-)

@lupyuen Thanks for reminder,  I think @yamt san just doesn't like the current constraints. In my eyes, he is a "reformist". He adheres to the standards while also being interested in new technical directions.

anchao commented:
I don't really care whether this PR can be merged or not. It has no impact on me. It was only when a student asked me what modlib's capabilities were that I realized it needed to be renamed. If someone is disgusted by this change, then just keep the status quo.

anchao commented:
> @anchao Is there a tiny possibility that we misread this? Because I see that we have support for the Breaking PR: https://lists.apache.org/thread/v7d7k1n2d1tl8n6o3pnocbz33knx3gmp
> 
> > i personally like the change itself and prefer loosening the rules in CONTRIBUTING.md though.
> 
> In any case: An objection to the Breaking PR should rightfully be accompanied by an Alternative Solution: https://www.apache.org/foundation/voting.html#Veto
> 
> > A -1 vote by a qualified voter stops a code-modification proposal in its tracks. This constitutes a veto, and it cannot be overruled nor overridden by anyone. Vetoes stand until and unless the individual withdraws their veto.
> 
> > To prevent vetoes from being used capriciously, the voter must provide with the veto a technical justification showing why the change is bad (opens a security exposure, negatively affects performance, etc. ). A veto without a justification is invalid and has no weight.

It doesn't matter, I have done what I should do and explained a lot

xiaoxiang781216 commented:
@anchao please rebase the patch, so we can merge it after your close the vote with approving.

lupyuen commented:
Sorry @xiaoxiang781216: We can't merge this Breaking PR because we have a Blocking Vote:
- https://github.com/apache/nuttx/pull/15765#issuecomment-2788825199

According to the Apache Rules for Code Changes: We need to resolve the Blocking Vote before merging the Breaking PR:
- https://www.apache.org/foundation/voting.html#Veto

anchao commented:
> @anchao please rebase the patch, so we can merge it after your close the vote with approving.

Done

anchao commented:
> Sorry @xiaoxiang781216: We can't merge this Breaking PR because we have a Blocking Vote:
> 
> * [[BREAKING]libc/elf: rename modlib to libelf #15765 (comment)](https://github.com/apache/nuttx/pull/15765#issuecomment-2788825199)
> 
> According to the Apache Rules for Code Changes: We need to resolve the Blocking Vote before merging the Breaking PR:
> 
> * https://www.apache.org/foundation/voting.html#Veto

I asked @yamt san for his opinion again in the email. Let's wait for reply.

lupyuen commented:
Thanks @anchao! The CI Jobs are still running non-stop, I'll check for errors here:
https://gist.github.com/nuttxpr

xiaoxiang781216 commented:
> Sorry @xiaoxiang781216: We can't merge this Breaking PR because we have a Blocking Vote:
> 
>     * [[BREAKING]libc/elf: rename modlib to libelf #15765 (comment)](https://github.com/apache/nuttx/pull/15765#issuecomment-2788825199)
> 
> 
> According to the Apache Rules for Code Changes: We need to resolve the Blocking Vote before merging the Breaking PR:
> 
>     * https://www.apache.org/foundation/voting.html#Veto

one person vote -1 shouldn't block the merge, otherwise the project will be dead.

xiaoxiang781216 commented:
@anchao please go  ahead to close vote with approving, I will merge this patch.

lupyuen commented:
@xiaoxiang781216: Unfortunately that's what our NuttX Community has agreed on. And it's also consistent with the Apache Way. Perhaps you would like to propose a change to the NuttX Contribution Rules:
https://github.com/apache/nuttx/pull/15950

xiaoxiang781216 commented:
> @xiaoxiang781216: Unfortunately that's what our NuttX Community has agreed on. And it's also consistent with the Apache Way. Perhaps you would like to propose a change to the NuttX Contribution Rules: #15950

1. One -1 vote from non PMC/commiter shouldn't block the breaking change.
2. If the number of "vote +1" exceeds a certain ratio of "vote -1" votes, the breaking change can be merged.

anchao commented:
> @xiaoxiang781216: Unfortunately that's what our NuttX Community has agreed on. And it's also consistent with the Apache Way. Perhaps you would like to propose a change to the NuttX Contribution Rules: #15950

@xiaoxiang781216 @lupyuen  

@yamt withdrew his -1 vote, I think we can go ahead

lupyuen commented:
@anchao Great! Let's wait for 72 hours before closing the vote, thanks for pushing this through :-)

lupyuen commented:
Sorry @anchao: Would you know why `esp32-devkitc:elf` is failing?
https://gist.github.com/lupyuen/4753bc001049f284dde7105d50726aa5
```
sudo docker run -it ghcr.io/apache/nuttx/apache-nuttx-ci-linux:latest /bin/bash -c '
  set -x ;
  uname -a ;
  cd ;
  pwd ;
  git clone https://github.com/anchao/nuttx nuttx --branch 25020501 ;
  git clone https://github.com/anchao/nuttx-apps apps --branch 25020501 ;
  pushd nuttx ; git reset --hard HEAD ; echo NuttX Source: https://github.com/apache/nuttx/tree/$(git rev-parse HEAD) ; popd ;
  pushd apps  ; git reset --hard HEAD ; echo NuttX Apps: https://github.com/apache/nuttx-apps/tree/$(git rev-parse HEAD) ; popd ;
  cd nuttx ;
  tools/configure.sh esp32-devkitc:elf ;
  make -j || echo '\''***** BUILD FAILED'\'' ;
'

xtensa-esp32-elf-ld: cannot open linker script file /root/nuttx/libs/libc/modlib/gnu-elf.ld: No such file or directory
```

anchao commented:
> ~Sorry @anchao: Would you know why `esp32-devkitc:elf` is failing?~ ~https://gist.github.com/lupyuen/4753bc001049f284dde7105d50726aa5~
> 
> ```
> sudo docker run -it ghcr.io/apache/nuttx/apache-nuttx-ci-linux:latest /bin/bash -c '
>   set -x ;
>   uname -a ;
>   cd ;
>   pwd ;
>   git clone https://github.com/anchao/nuttx nuttx --branch 25020501 ;
>   git clone https://github.com/anchao/nuttx-apps apps --branch 25020501 ;
>   pushd nuttx ; git reset --hard HEAD ; echo NuttX Source: https://github.com/apache/nuttx/tree/$(git rev-parse HEAD) ; popd ;
>   pushd apps  ; git reset --hard HEAD ; echo NuttX Apps: https://github.com/apache/nuttx-apps/tree/$(git rev-parse HEAD) ; popd ;
>   cd nuttx ;
>   tools/configure.sh esp32-devkitc:elf ;
>   make -j || echo '\''***** BUILD FAILED'\'' ;
> '
> 
> xtensa-esp32-elf-ld: cannot open linker script file /root/nuttx/libs/libc/modlib/gnu-elf.ld: No such file or directory
> ```
> 
> Update: The build is OK now: https://gist.github.com/lupyuen/764344525cd60840c2a06cfd06bf758b

@lupyuen Conflict with #16154, please help to  trigger again...
https://github.com/apache/nuttx/pull/16154

lupyuen commented:
Wonder if anyone has a faster Ubuntu PC that can run the script below and capture the CI Logs? Otherwise we'll be stuck waiting another 36 hours for my PC to complete the CI Checks  :-(
```bash
  for job in \
    arm-01 arm-02 arm-03 arm-04 \
    arm-05 arm-06 arm-07 arm-08 \
    arm-09 arm-10 arm-11 arm-12 \
    arm-13 arm-14 \
    arm64-01 \
    other \
    risc-v-01 risc-v-02 risc-v-03 risc-v-04 \
    risc-v-05 risc-v-06 risc-v-07 \
    sim-01 sim-02 sim-03 \
    x86_64-01 \
    xtensa-01 xtensa-02 xtensa-03
  do

    sudo docker run -it \
      ghcr.io/apache/nuttx/apache-nuttx-ci-linux:latest \
      /bin/bash -c "
      set -x ;
      uname -a ;
      cd ;
      pwd ;
      git clone https://github.com/anchao/nuttx nuttx --branch 25020501 ;
      git clone https://github.com/anchao/nuttx-apps apps --branch 25020501 ;
      pushd nuttx ; git reset --hard HEAD ; echo NuttX Source: https://github.com/apache/nuttx/tree/\$(git rev-parse HEAD) ; popd ;
      pushd apps  ; git reset --hard HEAD ; echo NuttX Apps: https://github.com/apache/nuttx-apps/tree/\$(git rev-parse HEAD) ; popd ;
      cd nuttx/tools/ci ;
      ( sleep 10800 ; echo Killing pytest after timeout... ; pkill -f pytest )&
      (./cibuild.sh -c -A -N -R testlist/$job.dat || echo '***** BUILD FAILED') ;    
      "
  done
```

anchao commented:
> Wonder if anyone has a faster Ubuntu PC that can run the script below and capture the CI Logs? Otherwise we'll be stuck waiting another 36 hours for my PC to complete the CI Checks :-(
> 
> ```shell
>   for job in \
>     arm-01 arm-02 arm-03 arm-04 \
>     arm-05 arm-06 arm-07 arm-08 \
>     arm-09 arm-10 arm-11 arm-12 \
>     arm-13 arm-14 \
>     arm64-01 \
>     other \
>     risc-v-01 risc-v-02 risc-v-03 risc-v-04 \
>     risc-v-05 risc-v-06 risc-v-07 \
>     sim-01 sim-02 sim-03 \
>     x86_64-01 \
>     xtensa-01 xtensa-02 xtensa-03
>   do
> 
>     sudo docker run -it \
>       ghcr.io/apache/nuttx/apache-nuttx-ci-linux:latest \
>       /bin/bash -c "
>       set -x ;
>       uname -a ;
>       cd ;
>       pwd ;
>       git clone https://github.com/anchao/nuttx nuttx --branch 25020501 ;
>       git clone https://github.com/anchao/nuttx-apps apps --branch 25020501 ;
>       pushd nuttx ; git reset --hard HEAD ; echo NuttX Source: https://github.com/apache/nuttx/tree/\$(git rev-parse HEAD) ; popd ;
>       pushd apps  ; git reset --hard HEAD ; echo NuttX Apps: https://github.com/apache/nuttx-apps/tree/\$(git rev-parse HEAD) ; popd ;
>       cd nuttx/tools/ci ;
>       ( sleep 10800 ; echo Killing pytest after timeout... ; pkill -f pytest )&
>       (./cibuild.sh -c -A -N -R testlist/$job.dat || echo '***** BUILD FAILED') ;    
>       "
>   done
> ```

Maybe we can just run the configuration if `CONFIG_ELF` enabled:

```
$ grep -nIr CONFIG_ELF=y boards/ | awk -F'/' '{print $4"/"$6}' 
xx3823/nsh
s698pm-dkit/nsh
s698pm-dkit/smp
xx3803/nsh
spresense/rndis_smp
spresense/module
spresense/wifi
spresense/wifi_smp
spresense/rndis
spresense/posix_spawn
spresense/rndis_composite
spresense/elf
nucleo-h743zi/elf
same70-xplained/pysim
olimex-stm32-p407/kelf
stm32f4discovery/composite
stm32f4discovery/wifi
stm32f4discovery/rndis
stm32f4discovery/posix_spawn
stm32f4discovery/adb
stm32f4discovery/elf
lm3s6965-ek/qemu-flat
lm3s6965-ek/qemu-protected
...
rv-virt/knsh64
arty_a7/knsh-tickless
arty_a7/knsh
esp32c3-devkit/elf
esp32-devkitc/elf
esp32s3-devkit/elf
avaota-a1/nsh
qemu-armv8a/knsh
```

e.g esp32-devkitc:elf that just failed is in this list

lupyuen commented:
Hi @anchao: Please remember to Close the Voting, so we don't receive any more Blocking Votes. Thanks :-)

I'm now running 6 Concurrent CI Jobs on my Ubuntu Server, hope to finish all CI Checks within 12 hours...

![Screenshot 2025-04-10 at 9 26 45 PM](https://github.com/user-attachments/assets/82bb18d8-184d-47dd-a8e0-5084efa8dbd7)

lupyuen reviewed (approved):
@anchao Please close the voting. Thanks :-)
This Breaking PR compiles with the [NuttX Breaking Changes Handling Process](https://github.com/apache/nuttx/blob/master/CONTRIBUTING.md#113-breaking-changes-handling-process):
- [CI Checks are successful](https://gist.github.com/search?q=user%3Anuttxpr+8b9d2c0cfc74361c212733fdcd119c9f6da628b4&ref=searchresults)
- Tested OK with OSTest on Real Hardware: [Avaota-A1 A527 Arm64 SBC](https://gist.github.com/lupyuen/ae0a726e84bb318d4dfad101836fdfe9) and [Oz64 SG2000 64-bit RISC-V SBC](https://gist.github.com/lupyuen/b8d23be87a9b29ec8e025002ec029de1)

anchao commented:
> @anchao Please close [the voting](https://lists.apache.org/thread/oh3cjvpfl0fz2rd3qlf034y6shcxwc0h). Thanks :-)
> 
> This Breaking PR compiles with the [NuttX Breaking Changes Handling Process](https://github.com/apache/nuttx/blob/master/CONTRIBUTING.md#113-breaking-changes-handling-process):
> 
> * [CI Checks are successful](https://gist.github.com/search?q=user%3Anuttxpr+8b9d2c0cfc74361c212733fdcd119c9f6da628b4&ref=searchresults)
> * Tested OK with OSTest on Real Hardware: [Avaota-A1 A527 Arm64 SBC](https://gist.github.com/lupyuen/ae0a726e84bb318d4dfad101836fdfe9) and [Oz64 SG2000 64-bit RISC-V SBC](https://gist.github.com/lupyuen/b8d23be87a9b29ec8e025002ec029de1)
> * [No Blocking Binding Vote](https://lists.apache.org/thread/oh3cjvpfl0fz2rd3qlf034y6shcxwc0h)

@lupyuen Thank you, I have sent the voting end email in the community group, the final result is:
```
+1: 9
+0: 3
-1: 0
```

tmedicci commented:
> Sorry @tmedicci: Is the build for `esp32s3-devkit:python` successful when we see these messages for our Special CI Build? https://gist.github.com/nuttxpr/ea0bcb70eec1207d7b83290535d4ea34#file-ci-xtensa-03-log-L183-L233
> 
> ```
> configure: WARNING: no system libmpdecimal found; falling back to bundled libmpdecimal (deprecated and scheduled for removal in Python 3.15)
>   WARNING: The scripts pip3 and pip3.13 are installed in '/root/apps/interpreters/python/install/host/bin' which is not on PATH.
>   Consider adding this directory to PATH or, if you prefer to suppress this warning, use --no-warn-script-location.
> WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager, possibly rendering your system unusable.It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv. Use the --root-user-action option if you know what you are doing and want to suppress this warning.
> configure: WARNING: using cross tools not prefixed with host triplet
> configure: WARNING: xtensa-esp32s3-nuttx/gcc is not supported
> /root/apps/interpreters/python/Python/configure: line 10553: test: too many arguments
> configure: WARNING: no system libmpdecimal found; falling back to bundled libmpdecimal (deprecated and scheduled for removal in Python 3.15)
> /root/apps/interpreters/python/Python/configure: line 27327: test: =: unary operator expected
> configure: WARNING:
> Platform "xtensa-esp32s3-nuttx" with compiler "gcc" is not supported by the
> CPython core team, see https://peps.python.org/pep-0011/ for more information.
> Created /root/apps/interpreters/python/build/target/lib/python313.zip (3.11 MiB)
>   [1/1] Normalize esp32s3-devkit/python
> Untracked files:
> 	libs/libc/elf_sys_symtab.c
> ```
> 
> The above messages have appeared since March 30 (probably earlier), so I assume they are OK? `testbuild.sh` is terminating with an error, probably because of the Untracked File. Thanks!
> 
> * https://gist.github.com/nuttxpr/556dd4a808060ed0ea9687888cbbb755#file-ci-xtensa-03-log-L179-L229

Sorry Lup, I completely missed your message. Just tested it: everything is fine with the build/runtime testing. Just wondering why it's generating this file (not ignored).
